### DISCUSSION TITLE
how to cumulate elements in array of objects

### DISCUSSION THREAD

**Discussion by quanguachong** [open-policy-agent/community#241] [OPA and Rego] [answered]
Input is an array of object, I want to generate an object by cumulating the array's elements.

Example Input: 

```json
{
  "containers": [
    {
          "cpu": "100m",
          "memory": "200Mi"
    },
    {
          "cpu": "25m",
          "memory": "400Mi"
    }
  ]
}

```

expect output:

```json
{
    "result": {
        "cpu": 125,
        "memory": 629145600000
    }
}
```

**Comment by quanguachong** [comment, marked as answer]
I resolve this issue with following code: 

```rego
result[key] := value {
    keys := { key | container := input.containers[i]; _ = container[key]}
    key := keys[_]
   # canonify_value is the function to convert string-value to number.
    values := [val | val := canonify_value(input.containers[_][key],key)]
    value := sum(values)
}
```

The core idea is convert  `input.containers` to the following object: 
```json
{
  "cpu": [100, 25],
  "memory": [209715200000, 419430400000]
}
```

and then use build-in function `sum` to sum elements of array.

Is there any better resolution to resolve such problem with rego?

**Reply by anderseknert** [reply]
Hi there! Did you see the `units.parse` [built-in function](https://www.openpolicyagent.org/docs/latest/policy-reference/#units)? It could probably replace the `canonify_value` one you're using there (or not, depending on what it does exactly).

Except for that, the approach seems reasonable to me 👍

**Reply by quanguachong** [reply]
thanks for reply.
